File: pyCombinatorial/web/js/algorithms/hgs.js
// ============================================================================
// hgs.js — Hybrid Genetic Search for the TSP
// Mirrors algorithm/hgs.py: population management with biased fitness,
// diversity control, ordered crossover, double-bridge mutation, and 2-opt.
// ============================================================================

import { distanceCalc } from '../core/distance.js';
import { mulberry32 } from './_shared.js';

export const meta = {
  id: 'hgs',
  label: 'Hybrid Genetic Search',
  category: 'Metaheuristic',
  description: 'Hybrid Genetic Search: diversity-aware population management, OX crossover, double-bridge mutation, and granular 2-opt local search.',
  params: [
    { key: 'populationSize',    label: 'Population',            type: 'int',   default: 25,   min: 4,    max: 300 },
    { key: 'offspringSize',     label: 'Offspring / generation', type: 'int',   default: 40,   min: 1,    max: 500 },
    { key: 'elite',             label: 'Elite',                 type: 'int',   default: 4,    min: 1,    max: 100 },
    { key: 'closest',           label: 'Closest for diversity', type: 'int',   default: 5,    min: 1,    max: 100 },
    { key: 'mutationRate',      label: 'Mutation rate',         type: 'float', default: 0.25, min: 0,    max: 1, step: 0.01 },
    { key: 'candidates',        label: '2-opt candidates',      type: 'int',   default: 20,   min: 1,    max: 100 },
    { key: 'localSearch',       label: 'Local search',          type: 'bool',  default: true },
    { key: 'localSearchPasses', label: '2-opt passes',          type: 'int',   default: 50,   min: 1,    max: 200 },
    { key: 'generations',       label: 'Generations',           type: 'int',   default: 500,  min: 1,    max: 5000 },
    { key: 'maxNoImprovement',  label: 'No-improvement stop',   type: 'int',   default: 150,  min: 1,    max: 5000 },
    { key: 'seed',              label: 'Random seed',           type: 'int',   default: 42,   min: 0,    max: 99999 },
  ],
};

function routeCost(distanceMatrix, route) {
  let distance = 0.0;
  const n = route.length;
  for (let i = 0; i < n; i++) {
    distance += distanceMatrix[route[i]][route[(i + 1) % n]];
  }
  return distance;
}

function toCityTour(route) {
  const tour = route.map(c => c + 1);
  tour.push(tour[0]);
  return tour;
}

function isSymmetric(distanceMatrix, atol = 1e-10) {
  const n = distanceMatrix.length;
  for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      if (Math.abs(distanceMatrix[i][j] - distanceMatrix[j][i]) > atol) return false;
    }
  }
  return true;
}

function candidateLists(distanceMatrix, candidates) {
  const n = distanceMatrix.length;
  const k = Math.max(1, Math.min(Math.floor(candidates), n - 1));
  const lists = [];
  for (let i = 0; i < n; i++) {
    const row = [];
    for (let j = 0; j < n; j++) {
      if (j !== i) row.push([j, distanceMatrix[i][j]]);
    }
    row.sort((a, b) => a[1] - b[1] || a[0] - b[0]);
    lists.push(row.slice(0, k).map(x => x[0]));
  }
  return lists;
}

function edgeSet(route, symmetric) {
  const n = route.length;
  const edges = new Set();
  for (let i = 0; i < n; i++) {
    let a = route[i];
    let b = route[(i + 1) % n];
    if (symmetric && a > b) [a, b] = [b, a];
    edges.add(`${a},${b}`);
  }
  return edges;
}

function brokenPairsDistance(edgesA, edgesB, n) {
  let common = 0;
  for (const e of edgesA) if (edgesB.has(e)) common++;
  return 1.0 - common / n;
}

function averageDistanceToClosest(individual, population, closest) {
  if (population.length <= 1) return 1.0;
  const distances = [];
  const n = individual.route.length;
  for (const other of population) {
    if (other === individual) continue;
    distances.push(brokenPairsDistance(individual.edges, other.edges, n));
  }
  distances.sort((a, b) => a - b);
  const k = Math.min(Math.floor(closest), distances.length);
  if (k <= 0) return 1.0;
  let sum = 0.0;
  for (let i = 0; i < k; i++) sum += distances[i];
  return sum / k;
}

function makeIndividual(distanceMatrix, route, symmetric) {
  const r = route.slice();
  return {
    route: r,
    cost: routeCost(distanceMatrix, r),
    edges: edgeSet(r, symmetric),
    fitness: 0.0,
  };
}

function normalizeRoute(route) {
  const r = route.slice();
  const p = r.indexOf(0);
  if (p >= 0) return r.slice(p).concat(r.slice(0, p));
  return r;
}

function randInt(rng, maxExclusive) {
  return Math.floor(rng() * maxExclusive);
}

function randomRoute(n, rng) {
  const route = Array.from({ length: n }, (_, i) => i);
  for (let i = n - 1; i > 0; i--) {
    const j = randInt(rng, i + 1);
    [route[i], route[j]] = [route[j], route[i]];
  }
  return normalizeRoute(route);
}

function randomizedNearestNeighbor(distanceMatrix, rng, restrictedCandidateList = 3) {
  const n = distanceMatrix.length;
  const start = randInt(rng, n);
  const unvisited = new Set(Array.from({ length: n }, (_, i) => i));
  unvisited.delete(start);
  const route = [start];
  let current = start;
  const rclSize = Math.max(1, Math.floor(restrictedCandidateList));

  while (unvisited.size > 0) {
    const ranked = [...unvisited].sort((a, b) => distanceMatrix[current][a] - distanceMatrix[current][b] || a - b);
    const limit = Math.min(rclSize, ranked.length);
    const next = ranked[randInt(rng, limit)];
    route.push(next);
    unvisited.delete(next);
    current = next;
  }
  return normalizeRoute(route);
}

function orderedCrossover(parentA, parentB, rng) {
  const n = parentA.length;
  let i = randInt(rng, n);
  let j = randInt(rng, n);
  while (i === j) j = randInt(rng, n);
  if (i > j) [i, j] = [j, i];

  const child = new Array(n).fill(-1);
  for (let k = i; k <= j; k++) child[k] = parentA[k];
  const used = new Set(child.slice(i, j + 1));
  let fillPos = (j + 1) % n;
  let scanPos = (j + 1) % n;

  for (let k = 0; k < n; k++) {
    const city = parentB[scanPos];
    if (!used.has(city)) {
      child[fillPos] = city;
      fillPos = (fillPos + 1) % n;
      used.add(city);
    }
    scanPos = (scanPos + 1) % n;
  }
  return normalizeRoute(child);
}

function swapMutation(route, rng) {
  const r = route.slice();
  const n = r.length;
  let i = randInt(rng, n);
  let j = randInt(rng, n);
  while (i === j) j = randInt(rng, n);
  [r[i], r[j]] = [r[j], r[i]];
  return normalizeRoute(r);
}

function doubleBridge(route, rng) {
  const n = route.length;
  if (n < 8) return swapMutation(route, rng);
  const cuts = new Set();
  while (cuts.size < 4) cuts.add(1 + randInt(rng, n - 1));
  const [a, b, c, d] = [...cuts].sort((x, y) => x - y);
  const newRoute = route.slice(0, a)
    .concat(route.slice(c, d), route.slice(b, c), route.slice(a, b), route.slice(d, n));
  return normalizeRoute(newRoute);
}

function twoOptLocalSearch(distanceMatrix, route, candidate, maxPasses = 50) {
  const n = route.length;
  const r = route.slice();
  let bestCost = routeCost(distanceMatrix, r);
  const eps = 1e-12;
  let passes = 0;
  let improved = true;

  while (improved && passes < maxPasses) {
    improved = false;
    passes++;
    const pos = new Array(n);
    for (let idx = 0; idx < n; idx++) pos[r[idx]] = idx;

    for (let i = 0; i < n - 1; i++) {
      const a = r[i];
      const b = r[(i + 1) % n];
      for (const c of candidate[a]) {
        const j = pos[c];
        if (j <= i + 1) continue;
        if (i === 0 && j === n - 1) continue;
        const d = r[(j + 1) % n];
        const delta = distanceMatrix[a][c] + distanceMatrix[b][d] - distanceMatrix[a][b] - distanceMatrix[c][d];
        if (delta < -eps) {
          const seg = r.slice(i + 1, j + 1).reverse();
          for (let k = 0; k < seg.length; k++) r[i + 1 + k] = seg[k];
          bestCost += delta;
          improved = true;
          break;
        }
      }
      if (improved) break;
    }
  }
  return [normalizeRoute(r), bestCost];
}

function updateBiasedFitness(population, elite, closest) {
  if (population.length === 0) return;
  population.sort((a, b) => a.cost - b.cost);
  const nPop = population.length;
  if (nPop === 1) {
    population[0].fitness = 0.0;
    return;
  }

  const diversity = population.map(ind => averageDistanceToClosest(ind, population, closest));
  const costOrder = Array.from({ length: nPop }, (_, i) => i).sort((a, b) => population[a].cost - population[b].cost);
  const divOrder = Array.from({ length: nPop }, (_, i) => i).sort((a, b) => diversity[b] - diversity[a]);
  const costRank = new Array(nPop).fill(0);
  const divRank = new Array(nPop).fill(0);

  for (let r = 0; r < nPop; r++) costRank[costOrder[r]] = r;
  for (let r = 0; r < nPop; r++) divRank[divOrder[r]] = r;

  const eliteCount = Math.max(1, Math.min(Math.floor(elite), nPop));
  for (let i = 0; i < nPop; i++) {
    const fitRank = costRank[i] / (nPop - 1);
    const divContributionRank = divRank[i] / (nPop - 1);
    population[i].fitness = nPop <= eliteCount
      ? fitRank
      : fitRank + (1.0 - eliteCount / nPop) * divContributionRank;
  }
}

function uniquePopulation(population) {
  const unique = new Map();
  for (const ind of population) {
    const key = ind.route.join(',');
    if (!unique.has(key) || ind.cost < unique.get(key).cost) unique.set(key, ind);
  }
  return [...unique.values()];
}

function survivorSelection(population, populationSize, elite, closest) {
  let pop = uniquePopulation(population);
  updateBiasedFitness(pop, elite, closest);
  while (pop.length > populationSize) {
    let worstIdx = 0;
    for (let i = 1; i < pop.length; i++) {
      if (pop[i].fitness > pop[worstIdx].fitness) worstIdx = i;
    }
    pop.splice(worstIdx, 1);
    updateBiasedFitness(pop, elite, closest);
  }
  pop.sort((a, b) => a.cost - b.cost);
  return pop;
}

function binaryTournament(population, rng) {
  const n = population.length;
  let a = randInt(rng, n);
  let b = randInt(rng, n);
  while (a === b) b = randInt(rng, n);
  return population[a].fitness <= population[b].fitness ? population[a] : population[b];
}

function cloneBest(individual) {
  return {
    route: individual.route.slice(),
    cost: individual.cost,
    edges: new Set(individual.edges),
    fitness: individual.fitness,
  };
}

export async function* run(distanceMatrix, params) {
  const n = distanceMatrix.length;
  const rng = mulberry32(params.seed ?? 42);
  const symmetric = isSymmetric(distanceMatrix);
  const populationSize = Math.max(4, Math.floor(params.populationSize ?? 25));
  const offspringSize = Math.max(1, Math.floor(params.offspringSize ?? 40));
  const elite = Math.max(1, Math.floor(params.elite ?? 4));
  const closest = Math.max(1, Math.floor(params.closest ?? 5));
  const mutationRate = Math.min(1, Math.max(0, Number(params.mutationRate ?? 0.25)));
  const candidates = Math.max(1, Math.floor(params.candidates ?? 20));
  const localSearch = params.localSearch !== false;
  const localSearchPasses = Math.max(1, Math.floor(params.localSearchPasses ?? 50));
  const generations = Math.max(1, Math.floor(params.generations ?? 500));
  const maxNoImprovement = Math.max(1, Math.floor(params.maxNoImprovement ?? 150));
  const candidate = candidateLists(distanceMatrix, candidates);

  let population = [];
  const initializationSize = Math.max(populationSize * 2, populationSize + offspringSize);

  for (let i = 0; i < initializationSize; i++) {
    let route = i < Math.floor(initializationSize / 2)
      ? randomizedNearestNeighbor(distanceMatrix, rng, 3)
      : randomRoute(n, rng);
    if (localSearch) [route] = twoOptLocalSearch(distanceMatrix, route, candidate, localSearchPasses);
    population.push(makeIndividual(distanceMatrix, route, symmetric));
  }

  population = survivorSelection(population, populationSize, elite, closest);
  let best = cloneBest(population[0]);
  let noImprovement = 0;
  let bestTour = toCityTour(best.route);
  let bestDistance = distanceCalc(distanceMatrix, bestTour);

  yield {
    phase: 'init',
    tour: bestTour,
    bestTour,
    distance: bestDistance,
    bestDistance,
    iteration: 0,
    message: `pop ${population.length}, start ${bestDistance.toFixed(2)}`,
  };

  for (let gen = 1; gen <= generations; gen++) {
    updateBiasedFitness(population, elite, closest);
    const offspring = [];

    for (let i = 0; i < offspringSize; i++) {
      const parentA = binaryTournament(population, rng);
      let parentB = binaryTournament(population, rng);
      let trials = 0;
      while (parentA === parentB && trials < 10) {
        parentB = binaryTournament(population, rng);
        trials++;
      }

      let childRoute = orderedCrossover(parentA.route, parentB.route, rng);
      if (rng() <= mutationRate) childRoute = doubleBridge(childRoute, rng);
      if (localSearch) [childRoute] = twoOptLocalSearch(distanceMatrix, childRoute, candidate, localSearchPasses);
      offspring.push(makeIndividual(distanceMatrix, childRoute, symmetric));
    }

    population = survivorSelection(population.concat(offspring), populationSize, elite, closest);

    const improved = population[0].cost < best.cost - 1e-12;
    if (improved) {
      best = cloneBest(population[0]);
      noImprovement = 0;
      bestTour = toCityTour(best.route);
      bestDistance = distanceCalc(distanceMatrix, bestTour);
    } else {
      noImprovement++;
    }

    const currentTour = toCityTour(population[0].route);
    const currentDistance = distanceCalc(distanceMatrix, currentTour);

    yield {
      phase: improved ? 'improvement' : 'generation',
      tour: currentTour,
      bestTour,
      distance: currentDistance,
      bestDistance,
      iteration: gen,
      message: improved ? `★ ${bestDistance.toFixed(2)}` : `gen ${gen}: ${currentDistance.toFixed(2)}  best ${bestDistance.toFixed(2)}`,
    };

    if (noImprovement >= maxNoImprovement) break;
  }

  return { tour: bestTour, distance: bestDistance, summary: `HGS: ${bestDistance.toFixed(2)}` };
}

File: pyCombinatorial/algorithm/hgs.py
############################################################################

# Created by: Prof. Valdecy Pereira, D.Sc.
# UFF - Universidade Federal Fluminense (Brazil)
# email:  valdecy.pereira@gmail.com
# Lesson: pyCombinatorial - Hybrid Genetic Search for the TSP
#
# GitHub Repository: <https://github.com/Valdecy>
#
# Notes:
#   This is a TSP-oriented Python adaptation of the general Hybrid Genetic
#   Search idea: population management with diversity control, ordered
#   crossover, mutation/perturbation, and local improvement. It is not a
#   direct CVRP port: the CVRP Split procedure, capacity penalties, duration
#   penalties, routes, and SWAP* inter-route neighborhood are intentionally
#   removed because a TSP solution is a single Hamiltonian cycle.
#
# Algorithm inspired by:
#   Vidal, T., Crainic, T. G., Gendreau, M., Lahrichi, N., Rei, W. (2012).
#   A hybrid genetic algorithm for multidepot and periodic vehicle routing
#   problems. Operations Research, 60(3), 611-624.
#
#   Vidal, T. (2022). Hybrid genetic search for the CVRP: Open-source
#   implementation and SWAP* neighborhood. Computers & Operations Research,
#   140, 105643.
#
############################################################################

# Required Libraries
import time
import random
import numpy as np

############################################################################

# Function: Tour Distance (same output convention as pyCombinatorial)
def distance_calc(distance_matrix, city_tour):
    distance = 0
    for k in range(0, len(city_tour[0]) - 1):
        m        = k + 1
        distance = distance + distance_matrix[city_tour[0][k] - 1, city_tour[0][m] - 1]
    return distance

############################################################################

# Function: Internal Closed-Tour Cost
def _route_cost(distance_matrix, route):
    n        = len(route)
    distance = 0.0
    for i in range(0, n):
        distance = distance + distance_matrix[route[i], route[(i + 1) % n]]
    return float(distance)

############################################################################

# Function: Convert Internal Route to pyCombinatorial City Tour
def _to_city_tour(distance_matrix, route):
    city_route = [int(i + 1) for i in route]
    city_route.append(city_route[0])
    distance   = distance_calc(distance_matrix, [city_route, 0])
    return [city_route, distance]

############################################################################

# Function: Matrix Check
def _prepare_distance_matrix(distance_matrix):
    distance_matrix = np.asarray(distance_matrix, dtype = float)
    if (distance_matrix.ndim != 2 or distance_matrix.shape[0] != distance_matrix.shape[1]):
        raise ValueError('distance_matrix must be a square matrix.')
    if (distance_matrix.shape[0] < 3):
        raise ValueError('TSP requires at least three cities.')
    if (not np.all(np.isfinite(distance_matrix))):
        raise ValueError('distance_matrix must contain only finite values.')
    return distance_matrix

############################################################################

# Function: Symmetry Check
def _is_symmetric(distance_matrix, atol = 1e-10):
    return bool(np.allclose(distance_matrix, distance_matrix.T, atol = atol, rtol = 0.0))

############################################################################

# Function: Nearest Candidate Lists
def _candidate_lists(distance_matrix, candidates):
    n          = distance_matrix.shape[0]
    candidates = int(max(1, min(candidates, n - 1)))
    cand       = []
    for i in range(0, n):
        row    = distance_matrix[i].copy()
        row[i] = np.inf
        cand.append(np.argsort(row, kind = 'stable')[:candidates].tolist())
    return cand

############################################################################

# Function: Tour Edges for Diversity Evaluation
def _edge_set(route, symmetric):
    n     = len(route)
    edges = set()
    for i in range(0, n):
        a = route[i]
        b = route[(i + 1) % n]
        if (symmetric == True and a > b):
            a, b = b, a
        edges.add((a, b))
    return edges

############################################################################

# Function: Broken-Pairs Distance
def _broken_pairs_distance(edges_a, edges_b, n):
    common = len(edges_a.intersection(edges_b))
    return 1.0 - common / float(n)

############################################################################

# Function: Average Distance to Closest Individuals
def _average_distance_to_closest(individual, population, closest):
    if (len(population) <= 1):
        return 1.0
    distances = []
    n         = len(individual['route'])
    for other in population:
        if (other is individual):
            continue
        distances.append(_broken_pairs_distance(individual['edges'], other['edges'], n))
    distances.sort()
    k = min(int(closest), len(distances))
    if (k <= 0):
        return 1.0
    return float(sum(distances[:k]) / k)

############################################################################

# Function: Build Individual
def _make_individual(distance_matrix, route, symmetric):
    route = list(route)
    return {
        'route'  : route,
        'cost'   : _route_cost(distance_matrix, route),
        'edges'  : _edge_set(route, symmetric),
        'fitness': 0.0,
    }

############################################################################

# Function: Normalize Route for Stable Representation
def _normalize_route(route):
    route = list(route)
    if (0 in route):
        p     = route.index(0)
        route = route[p:] + route[:p]
    return route

############################################################################

# Function: Random Route
def _random_route(n, rng):
    route = list(range(0, n))
    rng.shuffle(route)
    return _normalize_route(route)

############################################################################

# Function: Randomized Nearest Neighbor Route
def _randomized_nearest_neighbor(distance_matrix, rng, restricted_candidate_list = 3):
    n          = distance_matrix.shape[0]
    start      = rng.randrange(n)
    unvisited  = set(range(0, n))
    unvisited.remove(start)
    route      = [start]
    current    = start
    rcl_size   = max(1, int(restricted_candidate_list))
    while (len(unvisited) > 0):
        ranked = sorted(unvisited, key = lambda j: distance_matrix[current, j])
        limit  = min(rcl_size, len(ranked))
        nxt    = ranked[rng.randrange(limit)]
        route.append(nxt)
        unvisited.remove(nxt)
        current = nxt
    return _normalize_route(route)

############################################################################

# Function: Ordered Crossover (OX)
def _ordered_crossover(parent_a, parent_b, rng):
    n = len(parent_a)
    i = rng.randrange(0, n)
    j = rng.randrange(0, n)
    while (i == j):
        j = rng.randrange(0, n)
    if (i > j):
        i, j = j, i
    child          = [-1] * n
    child[i:j + 1] = parent_a[i:j + 1]
    used           = set(child[i:j + 1])
    fill_pos       = (j + 1) % n
    scan_pos       = (j + 1) % n
    for _ in range(0, n):
        city = parent_b[scan_pos]
        if (city not in used):
            child[fill_pos] = city
            fill_pos        = (fill_pos + 1) % n
            used.add(city)
        scan_pos = (scan_pos + 1) % n
    return _normalize_route(child)

############################################################################

# Function: Swap Mutation
def _swap_mutation(route, rng):
    n     = len(route)
    route = list(route)
    i, j  = rng.sample(range(0, n), 2)
    route[i], route[j] = route[j], route[i]
    return _normalize_route(route)

############################################################################

# Function: Double-Bridge Perturbation
def _double_bridge(route, rng):
    n = len(route)
    if (n < 8):
        return _swap_mutation(route, rng)
    cuts = sorted(rng.sample(range(1, n), 4))
    a, b, c, d = cuts
    new_route = route[0:a] + route[c:d] + route[b:c] + route[a:b] + route[d:n]
    return _normalize_route(new_route)

############################################################################

# Function: 2-Opt Local Search with Candidate Lists
def _two_opt_local_search(distance_matrix, route, candidate, max_passes = 50):
    n         = len(route)
    route     = list(route)
    best_cost = _route_cost(distance_matrix, route)
    eps       = 1e-12
    passes    = 0
    improved  = True
    while (improved == True and passes < max_passes):
        improved = False
        passes   = passes + 1
        pos      = [0] * n
        for idx, city in enumerate(route):
            pos[city] = idx
        for i in range(0, n - 1):
            a = route[i]
            b = route[(i + 1) % n]
            for c in candidate[a]:
                j = pos[c]
                if (j <= i + 1):
                    continue
                if (i == 0 and j == n - 1):
                    continue
                d     = route[(j + 1) % n]
                delta = (distance_matrix[a, c] + distance_matrix[b, d]
                         - distance_matrix[a, b] - distance_matrix[c, d])
                if (delta < -eps):
                    route[i + 1:j + 1] = reversed(route[i + 1:j + 1])
                    best_cost          = best_cost + float(delta)
                    improved           = True
                    break
            if (improved == True):
                break
    return _normalize_route(route), float(best_cost)

############################################################################

# Function: Update Biased Fitness
def _update_biased_fitness(population, elite, closest):
    if (len(population) == 0):
        return
    population.sort(key = lambda ind: ind['cost'])
    n_pop = len(population)
    if (n_pop == 1):
        population[0]['fitness'] = 0.0
        return

    diversity = []
    for ind in population:
        diversity.append(_average_distance_to_closest(ind, population, closest))

    cost_order = sorted(range(0, n_pop), key = lambda i: population[i]['cost'])
    div_order  = sorted(range(0, n_pop), key = lambda i: diversity[i], reverse = True)

    cost_rank = [0] * n_pop
    div_rank  = [0] * n_pop

    for r, idx in enumerate(cost_order):
        cost_rank[idx] = r
    for r, idx in enumerate(div_order):
        div_rank[idx] = r

    elite = max(1, min(int(elite), n_pop))
    for i in range(0, n_pop):
        fit_rank = cost_rank[i] / float(n_pop - 1)
        div_contribution_rank = div_rank[i] / float(n_pop - 1)
        if (n_pop <= elite):
            population[i]['fitness'] = fit_rank
        else:
            population[i]['fitness'] = fit_rank + (1.0 - elite / float(n_pop)) * div_contribution_rank

############################################################################

# Function: Remove Duplicate Tours
def _unique_population(population):
    unique = {}
    for ind in population:
        key = tuple(ind['route'])
        if (key not in unique or ind['cost'] < unique[key]['cost']):
            unique[key] = ind
    return list(unique.values())

############################################################################

# Function: Survivor Selection
def _survivor_selection(population, population_size, elite, closest):
    population = _unique_population(population)
    _update_biased_fitness(population, elite, closest)
    while (len(population) > population_size):
        worst_idx = max(range(0, len(population)), key = lambda i: population[i]['fitness'])
        del population[worst_idx]
        _update_biased_fitness(population, elite, closest)
    population.sort(key = lambda ind: ind['cost'])
    return population

############################################################################

# Function: Binary Tournament
def _binary_tournament(population, rng):
    a, b = rng.sample(range(0, len(population)), 2)
    if (population[a]['fitness'] <= population[b]['fitness']):
        return population[a]
    return population[b]

############################################################################

# Function: Stop Condition
def _time_exceeded(start_time, time_limit):
    if (time_limit is None):
        return False
    return (time.time() - start_time) >= float(time_limit)

############################################################################

# Function: HGS for TSP
def hybrid_genetic_search(distance_matrix, population_size = 25, offspring_size = 40,
                          elite = 4, closest = 5, mutation_rate = 0.25,
                          candidates = 20, local_search = True,
                          local_search_passes = 50, iterations = 500,
                          max_no_improvement = 150, time_limit = None,
                          seed = None, verbose = True):
    """
    Hybrid Genetic Search for the Traveling Salesman Problem.

    Parameters
    ----------
    distance_matrix : numpy.ndarray
        Square distance/cost matrix.
    population_size : int
        Number of individuals preserved after survivor selection.
    offspring_size : int
        Number of children generated per iteration.
    elite : int
        Number of elite individuals protected in biased-fitness ranking.
    closest : int
        Number of closest individuals used to estimate diversity contribution.
    mutation_rate : float
        Probability of applying double-bridge mutation to a child.
    candidates : int
        Number of nearest neighbors used by granular 2-opt local search.
    local_search : bool
        If True, applies 2-opt local improvement to initial individuals and children.
    local_search_passes : int
        Maximum number of 2-opt improvement passes per local search call.
    iterations : int
        Maximum number of HGS generations.
    max_no_improvement : int
        Stop after this many generations without improving the incumbent.
    time_limit : float or None
        Optional wall-clock time limit in seconds.
    seed : int or None
        Random seed.
    verbose : bool
        If True, prints progress.

    Returns
    -------
    route : list
        Closed tour using pyCombinatorial's 1-indexed convention.
    distance : float
        Total tour distance.
    """
    distance_matrix    = _prepare_distance_matrix(distance_matrix)
    n                  = distance_matrix.shape[0]
    rng                = random.Random(seed)
    symmetric          = _is_symmetric(distance_matrix)
    candidate          = _candidate_lists(distance_matrix, candidates)
    population_size    = max(4, int(population_size))
    offspring_size     = max(1, int(offspring_size))
    elite              = max(1, int(elite))
    closest            = max(1, int(closest))
    iterations         = max(1, int(iterations))
    max_no_improvement = max(1, int(max_no_improvement))
    mutation_rate      = min(1.0, max(0.0, float(mutation_rate)))
    start_time         = time.time()

    # Initial population: randomized nearest-neighbor plus random tours.
    population = []
    initialization_size = max(population_size * 2, population_size + offspring_size)
    for i in range(0, initialization_size):
        if (_time_exceeded(start_time, time_limit) == True):
            break
        if (i < initialization_size // 2):
            route = _randomized_nearest_neighbor(distance_matrix, rng, restricted_candidate_list = 3)
        else:
            route = _random_route(n, rng)
        if (local_search == True):
            route, _ = _two_opt_local_search(distance_matrix, route, candidate, local_search_passes)
        population.append(_make_individual(distance_matrix, route, symmetric))

    population = _survivor_selection(population, population_size, elite, closest)
    best       = dict(population[0])
    no_improv  = 0

    for it in range(1, iterations + 1):
        if (_time_exceeded(start_time, time_limit) == True):
            break
        _update_biased_fitness(population, elite, closest)
        offspring = []
        for _ in range(0, offspring_size):
            if (_time_exceeded(start_time, time_limit) == True):
                break
            parent_a = _binary_tournament(population, rng)
            parent_b = _binary_tournament(population, rng)
            trials   = 0
            while (parent_a is parent_b and trials < 10):
                parent_b = _binary_tournament(population, rng)
                trials   = trials + 1

            child_route = _ordered_crossover(parent_a['route'], parent_b['route'], rng)

            if (rng.random() <= mutation_rate):
                child_route = _double_bridge(child_route, rng)

            if (local_search == True):
                child_route, _ = _two_opt_local_search(distance_matrix, child_route, candidate, local_search_passes)

            offspring.append(_make_individual(distance_matrix, child_route, symmetric))

        population = _survivor_selection(population + offspring, population_size, elite, closest)

        if (population[0]['cost'] < best['cost'] - 1e-12):
            best      = dict(population[0])
            no_improv = 0
        else:
            no_improv = no_improv + 1

        if (verbose == True):
            print('Generation = ', it, 'Distance = ', round(best['cost'], 4))

        if (no_improv >= max_no_improvement):
            break

    route, distance = _to_city_tour(distance_matrix, best['route'])
    return route, distance

############################################################################
